src/stepFunctions/commands/visualizeStateMachine/aslVisualization.ts

```diff
@@ -153,16 +153,28 @@ export class AslVisualization {
         )
 
         // When the panel is closed, dispose of any disposables/remove subscriptions
-        panel.onDidDispose(() => {
+        const disposePanel = () => {
             debouncedUpdate.cancel()
             this.onVisualizationDisposeEmitter.fire()
             this.disposables.forEach(disposable => {
                 disposable.dispose()
             })
             this.onVisualizationDisposeEmitter.dispose()
             this.isPanelDisposed = true
+        }
+
+        panel.onDidDispose(() => {
+            if (!this.isPanelDisposed) {
+                disposePanel()
+            }
         })
 
+        panel.dispose = () => {
+            if (!this.isPanelDisposed) {
+                disposePanel()
```

Comment:
Technically you just need to set the flag here and then `onDidDispose` will do the work asynchronously. But this works as well.

The key was for the flag to be set here so that `sendUpdateMessage` would read the correct dispose state flag and immediately stop without trying to perform actions on the (possibly) disposed webview.

Reply:
Is there a scenario where onDidDispose runs into problems because it is executed after the panel is disposed? The current ordering ensures that the call to disposePanel returns before actually disposing of the panel.

Reply:
I think it's just an optimization. `onDidDipose` is async so any subscribers to that notification can clean up their resources in the background. Doing too much work in `dispose` itself is probably not advised.

Reply:
What are your thoughts on removing the override of onDidDispose and instead having the call to disposePanel come from the overridden dispose method? Would that fix the race condition as long as we set the isPanelDisposed flag? In the actual extension the dispose method is executed before the call to onDidDispose, so maybe this is a simpler fix?